test/e2e/app/app.go

```diff
@@ -305,6 +314,52 @@ func (app *Application) ApplySnapshotChunk(req abci.RequestApplySnapshotChunk) a
 }
 
 func (app *Application) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
+	var sum int64
+	var extCount int
+	for _, vote := range req.LocalLastCommit.Votes {
+		if !vote.SignedLastBlock || len(vote.VoteExtension) == 0 {
+			continue
+		}
+		extValue, errVal := binary.Varint(vote.VoteExtension)
+		// This should have been verified in VerifyVoteExtension
+		if errVal <= 0 {
+			panic(fmt.Sprintf("Failed to parse vote extension. Got return value %d", errVal))
+		}
+		valAddr := crypto.Address(vote.Validator.Address)
+		app.logger.Info("got vote extension value in PrepareProposal", "valAddr", valAddr, "value", extValue)
+		sum += extValue
+		extCount++
+	}
+	// We only generate our special transaction if we have vote extensions
+	if extCount > 0 {
+		extTxPrefix := fmt.Sprintf("%s=", voteExtensionKey)
+		extTx := []byte(fmt.Sprintf("%s%d", extTxPrefix, sum))
+		app.logger.Info("preparing proposal with custom transaction from vote extensions", "tx", extTx)
+		// Our generated transaction takes precedence over any supplied
+		// transaction that attempts to modify the "extensionSum" value.
+		txRecords := make([]*abci.TxRecord, len(req.Txs)+1)
+		for i, tx := range req.Txs {
```

Comment:
We need to make sure the total space taken by transactions is not greater than `req.MaxTxBytes`
(this is because of #8210).
I can also take care of this in #8210, if I merge after you. In that case, please add a TODO here

Reply:
This commit still applies

Reply:
I now check that our total bytes from all txs doesn't exceed `req.MaxTxBytes`, and if adding a vote extension transaction exceeds this limit we skip adding that tx.

I do, however, assume that Tendermint supplies us with an initial list of txs that don't exceed `req.MaxTxBytes`.

Reply:
Thanks for the change. But we still have a (small) problem here: the latest spec we agreed in last ABCI++ meeting does not "force" Tendermint to send less txs than max_bytes. IOW, Tendermint _could_ send up  a tx list that doesn't fit in a block (I can give you further details as to why).
Even if we know today's Tendermint code won't do it, it might happen in the future. So all Apps must make sure that the returned transaction records fit in the block, and MUST remove some otherwise.
Take a look at [this](https://github.com/tendermint/tendermint/blob/969690d/spec/abci%2B%2B/abci%2B%2B_methods_002_draft.md?plain=1#L320)

Reply:
So then what would "good behavior" look like here? Would we just return a proposal with as many txs as we can fit in, without explicitly removing any excess txs?

Reply:
Yes, that's the logic we implemented elsewhere in the code base. If you don't include the excess txs, they're just left in the mempool, as though Tendermint hadn't reaped them.
Take a loot at e.g. [here](https://github.com/tendermint/tendermint/blob/8df38db/abci/types/application.go#L97)